src/pdb/mod.rs

```diff
@@ -201,64 +135,60 @@ pub struct Header {
     unknown: u32,
     /// Always incremented by at least one, sometimes by two or three.
     pub sequence: u32,
+    /// The gap seems to be always zero.
+    #[br(temp, assert(gap == 0))]
+    #[bw(calc = 0u32)]
+    gap: u32,
     /// Each table is a linked list of pages containing rows of a particular type.
+    #[br(count = num_tables)]
     pub tables: Vec<Table>,
 }
 
 impl Header {
-    /// Parse the header of a PDB file.
-    pub fn parse(input: &[u8]) -> IResult<&[u8], Header> {
-        // Unknown purpose, perhaps an unoriginal signature, seems to always have the value 0.
-        let (input, _) = nom::bytes::complete::tag(b"\0\0\0\0")(input)?;
-
-        let (input, page_size) = nom::number::complete::le_u32(input)?;
-        let (input, num_tables) = nom::number::complete::le_u32(input)?;
-        let table_count = usize::try_from(num_tables)
-            .map_err(|_| nom_input_error_with_kind(input, ErrorKind::TooLarge))?;
-
-        let (input, next_unused_page) = PageIndex::parse(input)?;
-        let (input, unknown) = nom::number::complete::le_u32(input)?;
-        let (input, sequence) = nom::number::complete::le_u32(input)?;
-
-        // Gap
-        let (input, _) = nom::bytes::complete::tag(b"\0\0\0\0")(input)?;
-
-        // Tables
-        let (input, tables) = nom::multi::count(Table::parse, table_count)(input)?;
-
-        Ok((
-            input,
-            Header {
-                page_size,
-                next_unused_page,
-                sequence,
-                unknown,
-                tables,
-            },
-        ))
-    }
-
     /// Returns the offset for the given `page_index`, relative to the start of the PDB file.
-    pub fn page_offset(&self, PageIndex(page_index): &PageIndex) -> Result<usize, TryFromIntError> {
-        (page_index * self.page_size).try_into()
+    #[must_use]
+    pub fn page_offset(&self, PageIndex(page_index): &PageIndex) -> u64 {
+        (page_index * self.page_size).into()
     }
 
-    /// Parses and returns a page from the original data slice.
-    pub fn page<'a>(&self, input: &'a [u8], page_index: &PageIndex) -> IResult<&'a [u8], Page> {
-        let position = self
-            .page_offset(page_index)
-            .map_err(|_| nom_input_error_with_kind(input, ErrorKind::TooLarge))?;
-        let (data, page) = Page::parse(&input[position..])?;
-        Ok((data, page))
+    /// Returns pages for the given Table.
+    pub fn read_pages<R: Read + Seek>(
+        &self,
+        reader: &mut R,
+        ro: &ReadOptions,
+        args: (&PageIndex, &PageIndex),
+    ) -> BinResult<Vec<Page>> {
+        let (first_page, last_page) = args;
+
+        let mut pages = vec![];
+        let mut page_index = first_page.clone();
```

Comment:
`pages.push(page);` is a move, which means we cannot use the borrowed `page.next_page` in the next iteration of the loop, which is why we need to clone.

The line you pointed out the clone is only necessary because there would be a type mismatch between `&PageIndex` and `PageIndex`.

Reply:
Makes sense, thanks for pointing that out. All hail borrowck ;)